### PR TITLE
2408 cherry picked

### PR DESCRIPTION
**Pull request by janciesko**

### CODE REVIEW

File: containers/unit_tests/TestDualView.hpp [hunk 1]
```diff
@@ -211,3 +211,355 @@ TEST(TEST_CATEGORY, dualview_deep_copy) {
 }  // namespace Test
 
 #endif  // KOKKOS_TEST_UNORDERED_MAP_HPP
+/*
```
Comment (jeffmiles63) [inline containers/unit_tests/TestDualView.hpp]: Looks like your copy of the file got appended to davids...they need to be merged together ... impl stuff in impl, test functions at the bottom, etc.

File: containers/unit_tests/TestDualView.hpp [hunk 2]
```diff
@@ -211,3 +211,355 @@ TEST(TEST_CATEGORY, dualview_deep_copy) {
...
+      for (size_t j = 0; j < a.h_view.extent(1); ++j) {
+        a_h_sum += a.h_view(i, j);
+    }
+
+    //Check
+    ASSERT_EQ(a_h_sum, a_d_sum);
+    ASSERT_EQ(a_h_sum, a.extent(0) * a.extent(1));  
+
+    /* Covers case "Realloc on Device" */
+    a.modify_host();
```
Comment (jeffmiles63) [inline containers/unit_tests/TestDualView.hpp:235]: The shrink test should verify that data is preserved.  Data is preserved if the host mod number is higher than the device...

```C++
    a.modify_host();
    Kokkos::deep_copy(a.h_view, 2);
    a.resize( n / factor, m / factor);
    a.sync_device();
```

File: containers/unit_tests/TestDualView.hpp [hunk 3]
```diff
@@ -211,3 +211,355 @@ TEST(TEST_CATEGORY, dualview_deep_copy) {
...
+    // Check host view is synced as expected
+    a_h_sum = 0;
+    for (size_t i = 0; i < a.h_view.extent(0); ++i)
+      for (size_t j = 0; j < a.h_view.extent(1); ++j) {
+        a_h_sum += a.h_view(i, j);
+    }
+
+    //Check
+    ASSERT_EQ(a_h_sum, a_d_sum);
+    ASSERT_EQ(a_h_sum, a.extent(0) * a.extent(1));
```
Comment (jeffmiles63) [inline containers/unit_tests/TestDualView.hpp]: If you change the contents as suggested above you should also change the verification value

File: containers/unit_tests/TestDualView.hpp [hunk 4]
```diff
@@ -211,3 +211,355 @@ TEST(TEST_CATEGORY, dualview_deep_copy) {
...
+
+    // Check host view is synced as expected
+    a_h_sum = 0;
+    for (size_t i = 0; i < a.h_view.extent(0); ++i)
+      for (size_t j = 0; j < a.h_view.extent(1); ++j) {
+        a_h_sum += a.h_view(i, j);
+    }
+
+    //Check
+    ASSERT_EQ(a_h_sum, a_d_sum);
```
Comment (jeffmiles63) [inline containers/unit_tests/TestDualView.hpp:262]: Should also check that the extents match the new values

```C++
    ASSERT_EQ(a.extent(0), n/factor);
    ASSERT_EQ(a.extent(1), m/factor);
```

File: containers/unit_tests/TestDualView.hpp (hunk 1 above)
Comment (janciesko) [reply containers/unit_tests/TestDualView.hpp]: Yes. Fixed.

File: containers/unit_tests/TestDualView.hpp (hunk 4 above)
Comment (janciesko) [reply containers/unit_tests/TestDualView.hpp:262]: Ok. Fixed.

File: containers/unit_tests/TestDualView.hpp (hunk 2 above)
Comment (janciesko) [reply containers/unit_tests/TestDualView.hpp:235]: Ok. I have removed the the deep-copy initialization of the device after shrinking to test if host data has been preserved.

File: containers/unit_tests/TestDualView.hpp (hunk 3 above)
Comment (janciesko) [reply containers/unit_tests/TestDualView.hpp]: Resized host data is synced to device which should maintain the verification values equal.

**Review by dhollman — APPROVED** [review]
LGTM

**Pull request merged**